### PR TITLE
add ruff to pre-commit and replace black action with pre-commit

### PR DESCRIPTION
**Pull request by karmacoma-eth**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -54,3 +54,6 @@ select = [
 ignore = [
     "E501",  # line too long
 ]
+exclude = [
+    "tests/lib/**"
```
Comment (karmacoma-eth) [inline pyproject.toml:58]: we don't want to run ruff on solady or vyper stuff we may import in external libs

File: src/halmos/sevm.py [hunk 2]
```diff
@@ -414,6 +415,8 @@ def push(self, v: Word) -> None:
             self.stack.append(simplify(v))
 
     def pop(self) -> Word:
+        if not self.stack:
+            raise StackUnderflowError()
```
Comment (karmacoma-eth) [inline src/halmos/sevm.py:419]: this used to fail with the raw Python `IndexError`, and caught by `pytest.raises(Exception)`

ruff rightfully points out that catching `Exception` is not great, so we can finally address the todo and wrap the exception in the semantically correct `StackUnderflowError`

File: tests/conftest.py [hunk 3]
```diff
@@ -1,3 +1,8 @@
+# workaround for ruff removing fixture imports
+# they look unused because of dependency injection by pytest
+from test_fixtures import *  # noqa
```
Comment (karmacoma-eth) [inline tests/conftest.py:3]: a bit annoying to work around, relevant issue: https://github.com/astral-sh/ruff/issues/3295

File: tests/test_cli.py [hunk 4]
```diff
@@ -76,16 +78,17 @@ def test_decode_mixed_bytecode():
     assert contract[0] == EVM.PUSH20
     assert contract[27] == EVM.RETURN
     assert contract[28] == EVM.STOP  # past the end
-
-    contract.valid_jump_destinations() == set()
+    assert contract.valid_jump_destinations() == set()
```
Comment (karmacoma-eth) [inline tests/test_cli.py:81]: ruff pointed out that we were missing the assert 😅

File: tests/test_config.py [hunk 5]
```diff
@@ -35,7 +39,7 @@ def test_fresh_config_has_only_None_values():
 
 
 def test_default_config_immutable(config):
-    with pytest.raises(Exception):
+    with pytest.raises(dataclasses.FrozenInstanceError):
```
Comment (karmacoma-eth) [inline tests/test_config.py:42]: more specific exception type 👌

**Review by daejunpark — APPROVED** [review]

**Pull request merged**